### PR TITLE
GuardRails In ARO

### PR DESCRIPTION
**Pull request by yjst2012**
### Which issue this PR addresses:

[ADR doc](https://docs.google.com/document/d/1e4fyaKYKQAZ9Ske0qqCU7vD7KT82080FvQXDfNebmzk/edit#heading=h.i1hq6tedaz2) 
[ADO task](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/14397853) 

### What this PR does / why we need it:

It integrates gatekeeper and helpful policies to provide guard rails for ARO end users.

### Test plan for issue:

[A test plan](https://docs.google.com/document/d/1lfJrpP6NvDsl6wtIuw3gIFiNXoyQCiyQgeFczpWbdKY/edit#heading=h.epoi5jsn8v1a) is in progress

### Is there any documentation that needs to be updated for this PR?

Above mentioned docs may need some update if necessary

### Please note: 
this is a draft PR that aims to collect issues and problems, and some major overhaul may still happen, so debug print messages are not removed, please ignore them, thanks.

### CODE REVIEW

File: pkg/util/dynamichelper/dynamichelper.go [hunk 1]
```diff
@@ -92,6 +135,83 @@ func (dh *dynamicHelper) Ensure(ctx context.Context, objs ...kruntime.Object) er
...
+			return err
+		}
+		create = true
+	}
+	if create {
+		dh.log.Printf("Create %s", keyFunc(o.GroupVersionKind().GroupKind(), o.GetNamespace(), o.GetName()))
+		if _, err = dh.dynamicClient.Resource(*gvr).Namespace(o.obj.GetNamespace()).Create(ctx, &o.obj, metav1.CreateOptions{}); err != nil {
+			return err
+		}
+		return nil
```
Comment (facchettos) [inline pkg/util/dynamichelper/dynamichelper.go:157]:
```suggestion
		return dh.dynamicClient.Resource(*gvr).Namespace(o.obj.GetNamespace()).Create(ctx, &o.obj, metav1.CreateOptions{})
```

**Review by facchettos — CHANGES_REQUESTED** [review]
some questions, some simplifications, some typos

but it's looking good so far!

File: pkg/operator/controllers/guardrails/staticresources/gk_pod_security_policy.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+apiVersion: policy/v1beta1
+kind: PodSecurityPolicy
+metadata:
+  # annotations:
+  #   seccomp.security.alpha.kubernetes.io/allowedProfileNames: '*'
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/operator/controllers/guardrails/staticresources/gk_pod_security_policy.yaml:5]: this is required by gatekeeper to be installed on openshift, [the link](https://open-policy-agent.github.io/gatekeeper/website/docs/vendor-specific/#running-on-openshift-4x) is for the latest gatekeeper, I dont have it for 3.9 which is the version used in the PR.
as to why it is commented instead of removed, this is for the convenience of upgrading, this section will be easily recognised as intended instead of omitted.

File: pkg/util/dynamichelper/dynamichelper.go [hunk 3]
```diff
@@ -66,12 +76,38 @@ func New(log *logrus.Entry, restconfig *rest.Config) (Interface, error) {
...
-	gvr, err := dh.Resolve(groupKind, "")
+	return dh.EnsureDeletedGVR(ctx, groupKind, namespace, name, "")
+}
+func (dh *dynamicHelper) EnsureDeletedGVR(ctx context.Context, groupKind, namespace, name, optionalVersion string) error {
+	gvr, err := dh.resolve(groupKind, optionalVersion)
 	if err != nil {
 		return err
 	}
 
+	// gatekeeper policies is unstructured and should be deleted differently
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:101]: gatekeeper requires ConstraintTemplate and Constraint, the kind of Constraint is defined in the specific ConstraintTemplate file, which is basically impossible to include a predefined scheme for that, so has to be unmarshalled as unstructured obj like [we did before](https://github.com/Azure/ARO-RP/blob/master/pkg/frontend/adminactions/kubeactions.go#L131-L144).
the creation / deletion api is different from schemed obj, hence has to be separated.

File: pkg/util/dynamichelper/dynamichelper.go [hunk 4]
```diff
@@ -66,12 +76,38 @@ func New(log *logrus.Entry, restconfig *rest.Config) (Interface, error) {
...
+	if isKindUnstructured(groupKind) {
+		dh.log.Printf("Delete unstructured obj kind %s ns %s name %s version %s", groupKind, namespace, name, optionalVersion)
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:103]: this is to keep the same log style with existing logs like:
dh.log.Printf("Create %s", keyFunc(gvk.GroupKind(), acc.GetNamespace(), acc.GetName()))
dh.log.Printf("Update %s: %s", keyFunc(gvk.GroupKind(), acc.GetNamespace(), acc.GetName()), diff)

File: pkg/util/dynamichelper/dynamichelper.go [hunk 5]
```diff
@@ -289,3 +413,96 @@ func makeURLSegments(gvr *schema.GroupVersionResource, namespace, name string) (
...
+	if err == nil || kerrors.IsNotFound(err) || strings.Contains(strings.ToLower(err.Error()), "notfound") {
+		return true
+	}
+	return false
+}
+
+// mergeGK takes the existing (old) and desired (new) objects.  It compares them
+// to see if an update is necessary, fixes up the *old* object if needed, and
+// returns the difference for debugging purposes.
+// this is because that old objects were changed by gatekeeper binaries which have to be kept!
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:427]: to better understand it, some gatekeeper knowledge and guardrails design is required, basically the aim is just taking field we might change from the new config and merge it to back to the old config, as old config has been changed by gatekeeper binaries in the gatekeeper image, will try to revise it a bit, but still not confident I can explain it enough, if you have better phrasing feel free to comment.

File: cmd/aro/operator.go [hunk 6]
```diff
@@ -232,6 +233,9 @@ func operator(ctx context.Context, log *logrus.Entry) error {
 			arocli, operatorcli)).SetupWithManager(mgr); err != nil {
 			return fmt.Errorf("unable to create controller %s: %v", ingress.ControllerName, err)
 		}
+		if err = (guardrails.NewReconciler(arocli, kubernetescli, dh)).SetupWithManager(mgr); err != nil {
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply cmd/aro/operator.go:236]: good catch

File: pkg/util/dynamichelper/unstructuredobj.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+func (o UnstructuredObj) GetObjectKind() schema.ObjectKind {
+	return o.obj.GetObjectKind()
+}
+
+func (o UnstructuredObj) DeepCopyObject() kruntime.Object {
+	if un := o.obj.DeepCopy(); un != nil {
+		out := NewUnstructuredObj(*un)
+		return out
+	}
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go:31]: haha, initially there was debug output, which was then removed but the code has not been adapted, will fix it, thanks

File: pkg/util/dynamichelper/unstructuredobj.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+func (o UnstructuredObj) GroupVersionKind() schema.GroupVersionKind {
+	return o.obj.GroupVersionKind()
+}
+
+func (o UnstructuredObj) GetNamespace() string {
+	return o.obj.GetNamespace()
+}
+func (o UnstructuredObj) GetName() string {
+	return o.obj.GetName()
+}
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go:44]: to provide these common methods for the caller

File: pkg/util/dynamichelper/unstructuredobj.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+func (o *UnstructuredObj) DecodeUnstructured(data []byte) error {
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go:46]: actually I wrote it in your suggested way and then changed to the current way for better encapsulation

File: pkg/util/dynamichelper/unstructuredobj.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	json, err := yaml.YAMLToJSON(data)
+	if err != nil {
+		return err
+	}
+	err = o.obj.UnmarshalJSON(json)
+	if err != nil {
+		return err
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go:55]: same effect after debug output line removed

File: pkg/util/dynamichelper/dynamichelper.go [hunk 11]
```diff
@@ -289,3 +413,96 @@ func makeURLSegments(gvr *schema.GroupVersionResource, namespace, name string) (
...
+			}
+			if cmpAndCopy(&nc.Resources.Limits, &ec.Resources.Limits) {
+				changed = true
+			}
+			if cmpAndCopy(&nc.Resources.Requests, &ec.Resources.Requests) {
+				changed = true
+			}
+		}
+	case *admissionregistrationv1.ValidatingWebhookConfiguration:
+		new, expect := new.(*admissionregistrationv1.ValidatingWebhookConfiguration), expect.(*admissionregistrationv1.ValidatingWebhookConfiguration)
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:457]: for the webhook cases, will separate it into 2 lines, as single line is quite long, but for the deployment case, feel like single line assignment is more clear and readable as it provide more contrast.

File: pkg/util/dynamichelper/dynamichelper.go [hunk 12]
```diff
@@ -120,13 +240,17 @@ func (dh *dynamicHelper) ensureOne(ctx context.Context, new kruntime.Object) err
 			return err
 		}
 
-		new, changed, diff, err := merge(old, new)
+		mergeFunc := merge
+		if strings.HasPrefix(acc.GetName(), "gatekeeper") ||
+			strings.HasPrefix(gvk.GroupKind().String(), "ConstraintTemplate.templates.gatekeeper") {
+			mergeFunc = mergeGK
+		}
+		candidate, changed, diff, err := mergeFunc(old, new)
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:248]: what do you suggest? it is still a func var no matter put the block into a func or not?
feel like a func var makes code more readable, isn't it?
I initially implemented it using a switch case block, which is long and unclear.

File: pkg/util/dynamichelper/dynamichelper.go [hunk 13]
```diff
@@ -92,6 +135,83 @@ func (dh *dynamicHelper) Ensure(ctx context.Context, objs ...kruntime.Object) er
...
+
+func (dh *dynamicHelper) deleteUnstructuredObj(ctx context.Context, groupKind, namespace, name string) error {
+	gvr, err := dh.resolve(groupKind, "")
+	if err != nil {
+		return err
+	}
+	if err = dh.dynamicClient.Resource(*gvr).Namespace(namespace).Delete(ctx, name, metav1.DeleteOptions{}); err != nil && !notFound(err) {
+		return err
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:212]: the logic is changed in your suggested code, which makes the first and second block in the func detects different things, feel like it is not good enough to convene me, lol

File: pkg/util/dynamichelper/dynamichelper.go (hunk 1 above)
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:157]: your suggested code does not even compile as the func only one return value:
```
too many arguments to return
	have (*unstructured.Unstructured, error)
	want (error)
```

File: pkg/util/deployer/deployer.go [hunk 14]
```diff
@@ -103,19 +103,34 @@ func (depl *deployer) Remove(ctx context.Context, data interface{}) error {
...
+		if getName, getNs := reflect.ValueOf(obj).MethodByName("GetName"),
+			reflect.ValueOf(obj).MethodByName("GetNamespace"); getName != reflect.ValueOf(nil) && getNs != reflect.ValueOf(nil) {
+			name := getName.Call(nil)
+			ns := getNs.Call(nil)
+			if reflect.TypeOf(obj).String() == "*v1.Namespace" {
+				// dont delete the namespace for now
+				namespaceName = name[0].String()
+				continue
+			}
+			err := depl.dh.EnsureDeletedGVR(ctx, obj.GetObjectKind().GroupVersionKind().GroupKind().String(), ns[0].String(), name[0].String(), "")
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/deployer/deployer.go:118]: thanks for pointing out this, using reflect func call is probably not very safe, changed it to directly extract the Name and Namespace field after validation check, which should be safe enough for the situation.

File: pkg/operator/controllers/guardrails/staticresources/gk_pod_security_policy.yaml [hunk 15]
```diff
@@ -0,0 +1,32 @@
+apiVersion: policy/v1beta1
+kind: PodSecurityPolicy
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/operator/controllers/guardrails/staticresources/gk_pod_security_policy.yaml:2]: the current yaml files is for gatekeeper 3.9, which is tagetting an previous kubernetes version, currently working on upgrading gatekeeper version, which should have this stuff upgraded.

File: pkg/util/dynamichelper/dynamichelper.go [hunk 16]
```diff
@@ -289,3 +413,96 @@ func makeURLSegments(gvr *schema.GroupVersionResource, namespace, name string) (
...
+				changed = true
+			}
+			if expect.Webhooks[i].TimeoutSeconds != nil &&
+				new.Webhooks[i].TimeoutSeconds != nil &&
+				*expect.Webhooks[i].TimeoutSeconds != *new.Webhooks[i].TimeoutSeconds {
+				*expect.Webhooks[i].TimeoutSeconds = *new.Webhooks[i].TimeoutSeconds
+				changed = true
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:488]: code moved

File: cmd/aro/operator.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply cmd/aro/operator.go]: we can fix the others in a following pr

File: pkg/util/dynamichelper/dynamichelper.go (hunk 1 above)
Comment (facchettos) [reply pkg/util/dynamichelper/dynamichelper.go]: ah, yes indeed, missed that

File: pkg/util/dynamichelper/dynamichelper.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/dynamichelper.go:212]: there is no logic change, these are equivalent blocks. If you absolutely want to keep the `return err` first, use the following statement. `&&`s are really hard to follow when you have negations on each side. 
```suggestion
	if err = dh.dynamicClient.Resource(*gvr).Namespace(namespace).Delete(ctx, name, metav1.DeleteOptions{}); !(err == nil || notFound(err)) {
		return err
	}
	return nil
```

File: pkg/util/dynamichelper/dynamichelper.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/dynamichelper.go:248]: I was suggesting something like (pseudo code but this is just to give you an idea)
```go
func mergeWithLogic(acc, gvk, old, new) (candidate, changed, diff, err){
    if strings.HasPrefix().....{
        return mergeGK(old,new)
    }
    return merge(old, new)
}
```

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go:43]: this wrapper seems unnecessary tbh. The only reason you need it here is because you wanted to use a method on the struct to decode it, but this is not even needed. I'd rather we just use the actual object instead of the wrapper

File: pkg/util/dynamichelper/dynamichelper.go [hunk 17]
```diff
@@ -438,65 +439,83 @@ func mergeGK(old, new kruntime.Object) (kruntime.Object, bool, string, error) {
...
 	}
-
 	return expect, changed, diff, nil
 }
 
-func cmpAndCopy(srcP, dstP *corev1.ResourceList) bool {
-	src, dst := *srcP, *dstP
+func handleDeployment(new, expect kruntime.Object) bool {
+	changed := false
+	newDeployment, expectDeployment := new.(*appsv1.Deployment), expect.(*appsv1.Deployment)
```
Comment (facchettos) [inline pkg/util/dynamichelper/dynamichelper.go:457]:
```suggestion
	newDeployment, expectedDeployment := new.(*appsv1.Deployment), expected.(*appsv1.Deployment)
```

Comment (facchettos) [reply pkg/util/dynamichelper/dynamichelper.go:457]: btw, shouldn't we be checking if the type casting is successful? otherwise we might end up with panics? (here, but also in the webhook cases)

File: pkg/util/dynamichelper/dynamichelper.go [hunk 18]
```diff
@@ -438,65 +439,83 @@ func mergeGK(old, new kruntime.Object) (kruntime.Object, bool, string, error) {
...
+		}
+		if expectWebhook.Webhooks[i].TimeoutSeconds != nil &&
+			newWebhook.Webhooks[i].TimeoutSeconds != nil &&
+			*expectWebhook.Webhooks[i].TimeoutSeconds != *newWebhook.Webhooks[i].TimeoutSeconds {
+			*expectWebhook.Webhooks[i].TimeoutSeconds = *newWebhook.Webhooks[i].TimeoutSeconds
+			changed = true
+		}
+	}
+	return changed
+}
```
Comment (facchettos) [inline pkg/util/dynamichelper/dynamichelper.go:515]:
```suggestion
func handleMutatingWebhook(new, expect kruntime.Object) bool {
	newWebhook := new.(*admissionregistrationv1.MutatingWebhookConfiguration)
	expectWebhook := expect.(*admissionregistrationv1.MutatingWebhookConfiguration)
	for i := range expectWebhook.Webhooks {
		if expectWebhook.Webhooks[i].FailurePolicy != nil &&
			newWebhook.Webhooks[i].FailurePolicy != nil &&
			*expectWebhook.Webhooks[i].FailurePolicy != *newWebhook.Webhooks[i].FailurePolicy {
			*expectWebhook.Webhooks[i].FailurePolicy = *newWebhook.Webhooks[i].FailurePolicy
			return true
		}
		if expectWebhook.Webhooks[i].TimeoutSeconds != nil &&
			newWebhook.Webhooks[i].TimeoutSeconds != nil &&
			*expectWebhook.Webhooks[i].TimeoutSeconds != *newWebhook.Webhooks[i].TimeoutSeconds {
			*expectWebhook.Webhooks[i].TimeoutSeconds = *newWebhook.Webhooks[i].TimeoutSeconds
			return true
		}
	}
	return false
}
```

Comment (facchettos) [reply pkg/util/dynamichelper/dynamichelper.go:515]: also maybe the naming isn't optimal, what do you think of `mutatingHookHasChanged` or something?

File: pkg/util/deployer/deployer.go [hunk 19]
```diff
@@ -103,19 +103,41 @@ func (depl *deployer) Remove(ctx context.Context, data interface{}) error {
...
+		errDelete := depl.dh.EnsureDeletedGVR(ctx, obj.GetObjectKind().GroupVersionKind().GroupKind().String(), ns, name, "")
+		if errDelete != nil {
+			errs = append(errs, errDelete)
+		}
+	}
+	if namespaceName != "" {
+		// remove the namespace
+		err := depl.dh.EnsureDeleted(ctx, "Namespace", "", namespaceName)
+		if err != nil {
+			errs = append(errs, err)
```
Comment (facchettos) [inline pkg/util/deployer/deployer.go:135]: could we have this in a function? would avoid having all these continue

File: pkg/util/dynamichelper/dynamichelper.go (hunk 17 above)
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:457]: the cast cannot fail as it is checked just before the func call:
`	case *appsv1.Deployment:
		changed = handleDeployment(new, expected)`

File: pkg/util/dynamichelper/dynamichelper.go (hunk 18 above)
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:515]: **hasChanged** is enough, combined with the func name and context, otherwise it is too long.

Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:515]: cannot return in the middle, the loop has to be finished, otherwise some changes may be lost.

File: pkg/util/dynamichelper/dynamichelper.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go:248]: I can make the change if you insist, but feel like the origin code is more clear and readable tbh, func pointer isn't a evil, its just like any other variable.

File: pkg/util/deployer/deployer.go (hunk 19 above)
Comment (yjst2012) [reply pkg/util/deployer/deployer.go:135]: sure

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go:43]: they are also called [here](https://github.com/Azure/ARO-RP/pull/2555/files#diff-4dc999c394ba7fdd74cd881c96667ce6b8d1a88c5f2d5872c198e57708cc9f53R114) by guardrails controller

File: pkg/util/dynamichelper/dynamichelper.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/dynamichelper.go]: done

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go]: done

File: pkg/util/dynamichelper/dynamichelper.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/dynamichelper.go:248]: in some cases it makes sense to use that, just not here

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go]: I know, but there isn't any value added compared to using the actual `unstructured.Unstructured`

[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go]: feel like there is some misunderstanding here, the reason of introducing *UnstructuredObj* is that unstructured.Unstructured is not a *kruntime.Object* and hence cannot be put into the []kruntime.Object as a parameter for many APIs.

[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go]:
```go
	GetObjectKind() schema.ObjectKind
	DeepCopyObject() Object
```
but then only these are required, not the others

File: pkg/util/dynamichelper/unstructuredobj.go [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type UnstructuredObj struct {
+	obj unstructured.Unstructured
+}
+
+func NewUnstructuredObj(obj *unstructured.Unstructured) *UnstructuredObj {
+	return &UnstructuredObj{*obj}
+}
```
Comment (facchettos) [inline pkg/util/dynamichelper/unstructuredobj.go:21]: Is this really useful? seems like it's an unnecessary wrapping

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go]: also, if you could add a comment saying why we have those methods that'd be great

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 20 above)
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go:21]: and also, why did you change all the methods to use pointers in your last commit?

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go]: they get called outside dh in guardrails pkg, this is why the methods were added:
> err = r.dh.EnsureDeletedGVR(ctx, uns.GroupVersionKind().GroupKind().String(), uns.GetNamespace(), uns.GetName(), uns.GroupVersionKind().Version)
these methods are required by EnsureDeletedGVR()

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 20 above)
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go:21]: feel better to have a construction func first.

it is because of the previous change, where the obj to be added to the []kruntime.Object changed from UnstructuredObj to *UnstructuredObj for better readability, so just let *UnstructuredObj be a kruntime.Object here.

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go]: I have seen that, but it is still not useful. at that point you just need the `unstructured.Unstructured` object, and you can just use `append(creates,uns.Unstructure{obj})` when you need it. 

this should just return an `unstructured.Unstructured` object (with the signature becoming `func DecodeUnstructured(data []byte) (unstructed.Unstructured, error) {`)
```go
func (o *UnstructuredObj) DecodeUnstructured(data []byte) error {
	json, err := yaml.YAMLToJSON(data)
	if err != nil {
		return unstructured.Unstructured{},  err
	}
        obj := &unstructured.Unstructured{}
	err = obj.UnmarshalJSON(json)
        return obj, err
}
```

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 20 above)
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go:21]: it isn't useful.

File: pkg/util/dynamichelper/unstructuredobj.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go]: dont quite understand what you mean, do you want to completely remove the UnstructuredObj?
if not, you have to expose two confusing structs to callers, ie, guardrails, in order to decode one thing, which is UnstructuredObj and unstructured.Unstructured, do that sounds good?

in my implementation, only UnstructuredObj is exposed, callers dont need to know anything about the unstructured.Unstructured, everything related to it is encapsulated inside the package.

[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go]: the unstructuredObject struct should just be here to implement the interface you need. the rest is not really useful, adding wrappers like this are making things more confusing in the end

[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go]: with the current code structure, I don't see a reasonable way to remove these methods while achieve the same functionality, if you already have something in your mind, can you please let me know? thanks

[the remark this answers is not in the document]
Comment (facchettos) [reply pkg/util/dynamichelper/unstructuredobj.go]: I already did.

[the remark this answers is not in the document]
Comment (yjst2012) [reply pkg/util/dynamichelper/unstructuredobj.go]: ah, thanks, this is a bit tricky, I dont see other  languages have this kind of ### reverse inheritance

File: pkg/operator/controllers/guardrails/gktemplates/aro-deny-labels.yaml [hunk 21]
```diff
@@ -0,0 +1,56 @@
+apiVersion: templates.gatekeeper.sh/v1
+kind: ConstraintTemplate
+metadata:
+  name: arodenylabels
+  annotations:
+    description: >-
+      Prevents changes in resources that contain specified labels, with values
```
Comment (boranx) [inline pkg/operator/controllers/guardrails/gktemplates/aro-deny-labels.yaml:7]: very nice implementation! and nice to see `opa` will take place to establish a policy architecture overall, here's a small question: would we export this to an external repo that both osd/rosa and aro can call? Because it looks usable for osd/rosa as well

Comment (yjst2012) [reply pkg/operator/controllers/guardrails/gktemplates/aro-deny-labels.yaml:7]: AFAIK, osd/rosa has customised webhook implementation of similar features, so there is interests to migrate to gatekeeper? we sure can have some discussion if it is.

Comment (boranx) [reply pkg/operator/controllers/guardrails/gktemplates/aro-deny-labels.yaml:7]: From my point of view, yeah to be able to manage the policies at hand would bring us the flexibility to sync configurations for all the fleets without considering the underlying infrastructure and make sure they are in the same state in terms of policy governance

A security patch can be applied to all fleets instantly including aro-osd-rosa which would be a lot faster than trying to implement it in 2 different ways 

Although webhooks have their own advantages (like they are native, using go etc) it might be nice to have a second standard option amongst aro-rosa-osd etc

Comment (facchettos) [reply pkg/operator/controllers/guardrails/gktemplates/aro-deny-labels.yaml:7]: there is a plan to move away from the customized webhooks for OSD as far I know, PMs and such wanted to converge and use OPA
